x-pack/legacy/plugins/lens/public/xy_visualization_plugin/xy_expression.tsx

```diff
@@ -122,20 +124,64 @@ const MemoizedChart = React.memo(XYChart);
 
 export function XYChartReportable(props: XYChartRenderProps) {
   const [isReady, setIsReady] = useState(false);
+  const [hasError, setError] = useState(false);
 
   // It takes a cycle for the XY chart to render. This prevents
   // reporting from printing a blank chart placeholder.
   useEffect(() => {
     setIsReady(true);
+    if (props.handlers.done) {
+      props.handlers.done();
+    }
   }, []);
 
+  useEffect(() => {
+    if (hasError) {
+      // Clear error state when the config or data is updated
+      setError(false);
+    }
+  }, [props.data, props.args.layers]);
+
   return (
     <VisualizationContainer className="lnsXyExpression__container" isReady={isReady}>
-      <MemoizedChart {...props} />
+      <XYErrorBoundary
+        hasError={hasError}
+        onError={() => {
+          setError(true);
+        }}
+      >
+        <MemoizedChart {...props} />
+      </XYErrorBoundary>
     </VisualizationContainer>
   );
 }
 
+class XYErrorBoundary extends React.Component<{ hasError: boolean; onError: () => void }> {
+  constructor(props: { hasError: boolean; onError: () => void }) {
+    super(props);
+  }
+
+  componentDidCatch() {
+    this.props.onError();
+  }
+
+  render() {
+    if (this.props.hasError) {
+      return (
+        <div>
+          <EuiIcon type="alert" size="l" color="warning" />
+          <EuiText>
+            {i18n.translate('xpack.lens.xyVisualization.renderingErrorLabel', {
+              defaultMessage: 'Error rendering',
+            })}
```

Comment:
`componentDidCatch` provides the caught error. I wonder if we should have a "show details" link or something which, if clicked, shows the actual error message, etc?